{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "150d84cf",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-10-01T10:31:55.394646Z",
     "iopub.status.busy": "2021-10-01T10:31:55.393848Z",
     "iopub.status.idle": "2021-10-01T10:31:55.402993Z",
     "shell.execute_reply": "2021-10-01T10:31:55.401758Z",
     "shell.execute_reply.started": "2021-10-01T10:31:55.394573Z"
    }
   },
   "source": [
    "# Functional inference\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0a3bcc7b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## setup and import packages \n",
    "\n",
    "import os\n",
    "import pandas as pd\n",
    "import qiime2 as q2\n",
    "import requests\n",
    "\n",
    "from qiime2 import Visualization\n",
    "\n",
    "data_dir = 'Data'\n",
    "    \n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "210739be",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_ipath(ids: list, img_output_path: str, verbose: bool = False):\n",
    "    \"\"\"Fetches a enriched pathways map from iPATH3 for given IDs.\"\"\"\n",
    "    url = 'https://pathways.embl.de/mapping.cgi'\n",
    "    \n",
    "    # remove colon from EC names\n",
    "    if ':' in ids[0]:\n",
    "        ids = [x.replace(':', '') for x in ids]\n",
    "    \n",
    "    if verbose:\n",
    "        print(f'Fetching iPATH3 diagram for ids: {ids}')\n",
    "    params = {\n",
    "        'default_opacity': 0.6,\n",
    "        'export_type': 'svg',\n",
    "        'selection': '\\n'.join(ids)\n",
    "    }   \n",
    "    response = requests.get(url=url, params=params)\n",
    "    \n",
    "    with open(img_output_path, 'wb') as img:\n",
    "        img.write(response.content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1af03ced-2f86-4f4c-9d9f-09a9ba0a6059",
   "metadata": {},
   "outputs": [],
   "source": [
    "# path to the picrust2 conda environment\n",
    "picrust_env = '/opt/conda/envs/picrust2/bin'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed5ba7be",
   "metadata": {},
   "source": [
    "<a id='picrust'></a>\n",
    "## 1. Functional inference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "09f86b29",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved FeatureTable[Frequency] to: Data/picrust2_results/ko_metagenome.qza\n",
      "Saved FeatureTable[Frequency] to: Data/picrust2_results/ec_metagenome.qza\n",
      "Saved FeatureTable[Frequency] to: Data/picrust2_results/pathway_abundance.qza\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "QIIME is caching your current deployment for improved performance. This may take a few moments and should only happen once per deployment.\n"
     ]
    }
   ],
   "source": [
    "%%script env picrust_env=\"$picrust_env\" data_dir=\"$data_dir\" bash\n",
    "\n",
    "# append the env location to PATH so that qiime\n",
    "# can find all required executables\n",
    "export PATH=$picrust_env:$PATH\n",
    "\n",
    "$picrust_env/qiime picrust2 full-pipeline \\\n",
    "    --i-seq $data_dir/1-rep-seqs_bac.qza \\\n",
    "    --i-table $data_dir/1-feature-table_bac.qza \\\n",
    "    --output-dir $data_dir/picrust2_results \\\n",
    "    --p-placement-tool sepp \\\n",
    "    --p-threads 2 \\\n",
    "    --p-hsp-method pic \\\n",
    "    --p-max-nsti 2 "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86634b05",
   "metadata": {},
   "source": [
    "In the `picrust2_results` subdirectory, 3 artifacts as `FeatureTable[Frequency]` semantic type and hold abundance information for various functional features:\n",
    "\n",
    "1. KO metagenome - abundances of [KEGG](https://www.kegg.jp) orthologs\n",
    "2. EC metagenome - abundance of enzymes (you could look some of them up [here](https://www.brenda-enzymes.org)\n",
    "3. MetaCyc pathways - abundance of [MetaCyc](https://metacyc.org) pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "60ad1e2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = pd.read_csv(f'{data_dir}/0-metadata_bac.tsv', sep='\\t', header=0, index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "efa10199",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[33mQIIME is caching your current deployment for improved performance. This may take a few moments and should only happen once per deployment.\u001b[0m\n",
      "\u001b[32mSaved Visualization to: Data/0-metadata_bac.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "## look at the metadata to check the available categories\n",
    "\n",
    "! qiime metadata tabulate \\\n",
    "    --m-input-file $data_dir/0-metadata_bac.tsv \\\n",
    "    --o-visualization $data_dir/0-metadata_bac.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "4945c514",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2-archive-o3da9o1k')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: b0457538-92ce-4d3e-9f6b-70350b3de92b>"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load(f'{data_dir}/0-metadata_bac.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d27be17e",
   "metadata": {},
   "source": [
    "read in all three artifacts using QIIME 2 Python API - we can view them as DataFrames:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f211bd88",
   "metadata": {},
   "outputs": [],
   "source": [
    "## read all three artifacts using QIIME 2 Python API - we can view them as DataFrames\n",
    "\n",
    "ko = q2.Artifact.load(f'{data_dir}/picrust2_results/ko_metagenome.qza').view(pd.DataFrame)\n",
    "ec = q2.Artifact.load(f'{data_dir}/picrust2_results/ec_metagenome.qza').view(pd.DataFrame)\n",
    "pa = q2.Artifact.load(f'{data_dir}/picrust2_results/pathway_abundance.qza').view(pd.DataFrame)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d430188",
   "metadata": {},
   "source": [
    " Look at the contents of each of resulted tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "b558937f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>K00001</th>\n",
       "      <th>K00002</th>\n",
       "      <th>K00003</th>\n",
       "      <th>K00004</th>\n",
       "      <th>K00005</th>\n",
       "      <th>K00006</th>\n",
       "      <th>K00007</th>\n",
       "      <th>K00008</th>\n",
       "      <th>K00009</th>\n",
       "      <th>K00010</th>\n",
       "      <th>...</th>\n",
       "      <th>K19777</th>\n",
       "      <th>K19778</th>\n",
       "      <th>K19779</th>\n",
       "      <th>K19780</th>\n",
       "      <th>K19784</th>\n",
       "      <th>K19785</th>\n",
       "      <th>K19787</th>\n",
       "      <th>K19788</th>\n",
       "      <th>K19789</th>\n",
       "      <th>K19791</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERR1842970</th>\n",
       "      <td>15449.813969</td>\n",
       "      <td>971.486189</td>\n",
       "      <td>34528.116336</td>\n",
       "      <td>3535.764948</td>\n",
       "      <td>7600.251321</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>28.862673</td>\n",
       "      <td>20364.046310</td>\n",
       "      <td>3617.470286</td>\n",
       "      <td>7721.839866</td>\n",
       "      <td>...</td>\n",
       "      <td>70.163140</td>\n",
       "      <td>60.683262</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>3932.992737</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.023977</td>\n",
       "      <td>6.958197e-04</td>\n",
       "      <td>1371.012062</td>\n",
       "      <td>0.176941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR1842971</th>\n",
       "      <td>24337.732003</td>\n",
       "      <td>2321.649421</td>\n",
       "      <td>33233.880392</td>\n",
       "      <td>2981.170227</td>\n",
       "      <td>30825.547308</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>57.444589</td>\n",
       "      <td>22027.097231</td>\n",
       "      <td>3252.271406</td>\n",
       "      <td>8862.652184</td>\n",
       "      <td>...</td>\n",
       "      <td>80.010124</td>\n",
       "      <td>28.670483</td>\n",
       "      <td>1.010501e-43</td>\n",
       "      <td>1.019405e-43</td>\n",
       "      <td>2476.815498</td>\n",
       "      <td>5.420460e-130</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.002508e-59</td>\n",
       "      <td>1548.033978</td>\n",
       "      <td>0.000042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR1842972</th>\n",
       "      <td>14342.961169</td>\n",
       "      <td>960.408474</td>\n",
       "      <td>28959.808177</td>\n",
       "      <td>2610.467378</td>\n",
       "      <td>8702.058235</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>530.975065</td>\n",
       "      <td>18624.631287</td>\n",
       "      <td>4602.460834</td>\n",
       "      <td>8358.751898</td>\n",
       "      <td>...</td>\n",
       "      <td>45.435638</td>\n",
       "      <td>22.880090</td>\n",
       "      <td>2.698830e+00</td>\n",
       "      <td>1.855271e+00</td>\n",
       "      <td>2535.299760</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1616.010723</td>\n",
       "      <td>0.000159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR1842973</th>\n",
       "      <td>12017.160329</td>\n",
       "      <td>817.032836</td>\n",
       "      <td>26826.928342</td>\n",
       "      <td>1573.399263</td>\n",
       "      <td>5191.602392</td>\n",
       "      <td>0.005725</td>\n",
       "      <td>11.432426</td>\n",
       "      <td>10655.419661</td>\n",
       "      <td>1746.203962</td>\n",
       "      <td>4707.490882</td>\n",
       "      <td>...</td>\n",
       "      <td>6.763428</td>\n",
       "      <td>0.009178</td>\n",
       "      <td>9.207730e-49</td>\n",
       "      <td>9.288866e-49</td>\n",
       "      <td>2175.476313</td>\n",
       "      <td>1.010187e+01</td>\n",
       "      <td>0.005649</td>\n",
       "      <td>1.681534e-03</td>\n",
       "      <td>1354.902476</td>\n",
       "      <td>1.863283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR1842974</th>\n",
       "      <td>9622.984324</td>\n",
       "      <td>686.467483</td>\n",
       "      <td>20649.067462</td>\n",
       "      <td>1586.746095</td>\n",
       "      <td>8960.118044</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>181.980799</td>\n",
       "      <td>15223.869999</td>\n",
       "      <td>3589.761646</td>\n",
       "      <td>6050.617238</td>\n",
       "      <td>...</td>\n",
       "      <td>36.840059</td>\n",
       "      <td>82.187635</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2382.522254</td>\n",
       "      <td>4.130646e+00</td>\n",
       "      <td>0.001444</td>\n",
       "      <td>3.230048e-01</td>\n",
       "      <td>1583.781577</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 10535 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  K00001       K00002        K00003       K00004  \\\n",
       "ERR1842970  15449.813969   971.486189  34528.116336  3535.764948   \n",
       "ERR1842971  24337.732003  2321.649421  33233.880392  2981.170227   \n",
       "ERR1842972  14342.961169   960.408474  28959.808177  2610.467378   \n",
       "ERR1842973  12017.160329   817.032836  26826.928342  1573.399263   \n",
       "ERR1842974   9622.984324   686.467483  20649.067462  1586.746095   \n",
       "\n",
       "                  K00005    K00006      K00007        K00008       K00009  \\\n",
       "ERR1842970   7600.251321  0.000000   28.862673  20364.046310  3617.470286   \n",
       "ERR1842971  30825.547308  0.000000   57.444589  22027.097231  3252.271406   \n",
       "ERR1842972   8702.058235  0.000000  530.975065  18624.631287  4602.460834   \n",
       "ERR1842973   5191.602392  0.005725   11.432426  10655.419661  1746.203962   \n",
       "ERR1842974   8960.118044  0.000000  181.980799  15223.869999  3589.761646   \n",
       "\n",
       "                 K00010  ...     K19777     K19778        K19779  \\\n",
       "ERR1842970  7721.839866  ...  70.163140  60.683262  0.000000e+00   \n",
       "ERR1842971  8862.652184  ...  80.010124  28.670483  1.010501e-43   \n",
       "ERR1842972  8358.751898  ...  45.435638  22.880090  2.698830e+00   \n",
       "ERR1842973  4707.490882  ...   6.763428   0.009178  9.207730e-49   \n",
       "ERR1842974  6050.617238  ...  36.840059  82.187635  0.000000e+00   \n",
       "\n",
       "                  K19780       K19784         K19785    K19787        K19788  \\\n",
       "ERR1842970  0.000000e+00  3932.992737   0.000000e+00  0.023977  6.958197e-04   \n",
       "ERR1842971  1.019405e-43  2476.815498  5.420460e-130  0.000000  1.002508e-59   \n",
       "ERR1842972  1.855271e+00  2535.299760   0.000000e+00  0.000000  0.000000e+00   \n",
       "ERR1842973  9.288866e-49  2175.476313   1.010187e+01  0.005649  1.681534e-03   \n",
       "ERR1842974  0.000000e+00  2382.522254   4.130646e+00  0.001444  3.230048e-01   \n",
       "\n",
       "                 K19789    K19791  \n",
       "ERR1842970  1371.012062  0.176941  \n",
       "ERR1842971  1548.033978  0.000042  \n",
       "ERR1842972  1616.010723  0.000159  \n",
       "ERR1842973  1354.902476  1.863283  \n",
       "ERR1842974  1583.781577  0.000000  \n",
       "\n",
       "[5 rows x 10535 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ko.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "660cb300",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EC:1.1.1.1</th>\n",
       "      <th>EC:1.1.1.10</th>\n",
       "      <th>EC:1.1.1.100</th>\n",
       "      <th>EC:1.1.1.101</th>\n",
       "      <th>EC:1.1.1.102</th>\n",
       "      <th>EC:1.1.1.103</th>\n",
       "      <th>EC:1.1.1.105</th>\n",
       "      <th>EC:1.1.1.107</th>\n",
       "      <th>EC:1.1.1.108</th>\n",
       "      <th>EC:1.1.1.11</th>\n",
       "      <th>...</th>\n",
       "      <th>EC:6.4.1.8</th>\n",
       "      <th>EC:6.5.1.1</th>\n",
       "      <th>EC:6.5.1.2</th>\n",
       "      <th>EC:6.5.1.3</th>\n",
       "      <th>EC:6.5.1.4</th>\n",
       "      <th>EC:6.5.1.5</th>\n",
       "      <th>EC:6.5.1.6</th>\n",
       "      <th>EC:6.5.1.7</th>\n",
       "      <th>EC:6.6.1.1</th>\n",
       "      <th>EC:6.6.1.2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERR1842970</th>\n",
       "      <td>44650.950511</td>\n",
       "      <td>92.439958</td>\n",
       "      <td>143848.117824</td>\n",
       "      <td>1.964992e-08</td>\n",
       "      <td>54.075241</td>\n",
       "      <td>3875.947110</td>\n",
       "      <td>1.224978</td>\n",
       "      <td>47.145824</td>\n",
       "      <td>1362.852930</td>\n",
       "      <td>28.862673</td>\n",
       "      <td>...</td>\n",
       "      <td>2.495300</td>\n",
       "      <td>9946.704888</td>\n",
       "      <td>61545.437557</td>\n",
       "      <td>16559.981016</td>\n",
       "      <td>883.844762</td>\n",
       "      <td>165.586706</td>\n",
       "      <td>3531.382305</td>\n",
       "      <td>3531.382305</td>\n",
       "      <td>27650.130099</td>\n",
       "      <td>35385.555374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR1842971</th>\n",
       "      <td>62720.523941</td>\n",
       "      <td>148.041974</td>\n",
       "      <td>152384.212303</td>\n",
       "      <td>2.636290e-27</td>\n",
       "      <td>142.583849</td>\n",
       "      <td>3813.824669</td>\n",
       "      <td>2.718365</td>\n",
       "      <td>104.896437</td>\n",
       "      <td>1012.921180</td>\n",
       "      <td>57.444589</td>\n",
       "      <td>...</td>\n",
       "      <td>0.149111</td>\n",
       "      <td>5546.880825</td>\n",
       "      <td>66940.525222</td>\n",
       "      <td>18681.739160</td>\n",
       "      <td>480.549778</td>\n",
       "      <td>65.059259</td>\n",
       "      <td>2286.052633</td>\n",
       "      <td>2286.052633</td>\n",
       "      <td>15668.870066</td>\n",
       "      <td>26005.100543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR1842972</th>\n",
       "      <td>40443.884877</td>\n",
       "      <td>113.025173</td>\n",
       "      <td>132862.647734</td>\n",
       "      <td>4.749872e-27</td>\n",
       "      <td>29.413927</td>\n",
       "      <td>2925.557471</td>\n",
       "      <td>7.403809</td>\n",
       "      <td>69.521268</td>\n",
       "      <td>891.150555</td>\n",
       "      <td>530.975065</td>\n",
       "      <td>...</td>\n",
       "      <td>0.188714</td>\n",
       "      <td>5668.948858</td>\n",
       "      <td>56537.924620</td>\n",
       "      <td>17769.282922</td>\n",
       "      <td>541.048698</td>\n",
       "      <td>13.775866</td>\n",
       "      <td>2413.540227</td>\n",
       "      <td>2413.540227</td>\n",
       "      <td>20936.738288</td>\n",
       "      <td>29057.587202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR1842973</th>\n",
       "      <td>29594.155475</td>\n",
       "      <td>99.688765</td>\n",
       "      <td>97811.595899</td>\n",
       "      <td>1.140156e-27</td>\n",
       "      <td>106.125050</td>\n",
       "      <td>2776.188614</td>\n",
       "      <td>6.083980</td>\n",
       "      <td>41.562785</td>\n",
       "      <td>948.102312</td>\n",
       "      <td>11.432426</td>\n",
       "      <td>...</td>\n",
       "      <td>25.862977</td>\n",
       "      <td>5646.766696</td>\n",
       "      <td>43973.364626</td>\n",
       "      <td>10091.179565</td>\n",
       "      <td>507.044075</td>\n",
       "      <td>14.532473</td>\n",
       "      <td>2132.786636</td>\n",
       "      <td>2132.786636</td>\n",
       "      <td>22608.261560</td>\n",
       "      <td>24874.074773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR1842974</th>\n",
       "      <td>32820.767588</td>\n",
       "      <td>18.680208</td>\n",
       "      <td>107682.192564</td>\n",
       "      <td>9.848864e-09</td>\n",
       "      <td>138.062521</td>\n",
       "      <td>3982.605918</td>\n",
       "      <td>0.042111</td>\n",
       "      <td>35.311147</td>\n",
       "      <td>1030.913782</td>\n",
       "      <td>181.980799</td>\n",
       "      <td>...</td>\n",
       "      <td>0.067717</td>\n",
       "      <td>5104.755972</td>\n",
       "      <td>45196.590257</td>\n",
       "      <td>15645.466503</td>\n",
       "      <td>699.376339</td>\n",
       "      <td>8.515516</td>\n",
       "      <td>1651.802933</td>\n",
       "      <td>1651.802933</td>\n",
       "      <td>3281.599633</td>\n",
       "      <td>16280.321819</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 2911 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              EC:1.1.1.1  EC:1.1.1.10   EC:1.1.1.100  EC:1.1.1.101  \\\n",
       "ERR1842970  44650.950511    92.439958  143848.117824  1.964992e-08   \n",
       "ERR1842971  62720.523941   148.041974  152384.212303  2.636290e-27   \n",
       "ERR1842972  40443.884877   113.025173  132862.647734  4.749872e-27   \n",
       "ERR1842973  29594.155475    99.688765   97811.595899  1.140156e-27   \n",
       "ERR1842974  32820.767588    18.680208  107682.192564  9.848864e-09   \n",
       "\n",
       "            EC:1.1.1.102  EC:1.1.1.103  EC:1.1.1.105  EC:1.1.1.107  \\\n",
       "ERR1842970     54.075241   3875.947110      1.224978     47.145824   \n",
       "ERR1842971    142.583849   3813.824669      2.718365    104.896437   \n",
       "ERR1842972     29.413927   2925.557471      7.403809     69.521268   \n",
       "ERR1842973    106.125050   2776.188614      6.083980     41.562785   \n",
       "ERR1842974    138.062521   3982.605918      0.042111     35.311147   \n",
       "\n",
       "            EC:1.1.1.108  EC:1.1.1.11  ...  EC:6.4.1.8   EC:6.5.1.1  \\\n",
       "ERR1842970   1362.852930    28.862673  ...    2.495300  9946.704888   \n",
       "ERR1842971   1012.921180    57.444589  ...    0.149111  5546.880825   \n",
       "ERR1842972    891.150555   530.975065  ...    0.188714  5668.948858   \n",
       "ERR1842973    948.102312    11.432426  ...   25.862977  5646.766696   \n",
       "ERR1842974   1030.913782   181.980799  ...    0.067717  5104.755972   \n",
       "\n",
       "              EC:6.5.1.2    EC:6.5.1.3  EC:6.5.1.4  EC:6.5.1.5   EC:6.5.1.6  \\\n",
       "ERR1842970  61545.437557  16559.981016  883.844762  165.586706  3531.382305   \n",
       "ERR1842971  66940.525222  18681.739160  480.549778   65.059259  2286.052633   \n",
       "ERR1842972  56537.924620  17769.282922  541.048698   13.775866  2413.540227   \n",
       "ERR1842973  43973.364626  10091.179565  507.044075   14.532473  2132.786636   \n",
       "ERR1842974  45196.590257  15645.466503  699.376339    8.515516  1651.802933   \n",
       "\n",
       "             EC:6.5.1.7    EC:6.6.1.1    EC:6.6.1.2  \n",
       "ERR1842970  3531.382305  27650.130099  35385.555374  \n",
       "ERR1842971  2286.052633  15668.870066  26005.100543  \n",
       "ERR1842972  2413.540227  20936.738288  29057.587202  \n",
       "ERR1842973  2132.786636  22608.261560  24874.074773  \n",
       "ERR1842974  1651.802933   3281.599633  16280.321819  \n",
       "\n",
       "[5 rows x 2911 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ec.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ccecd600",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1CMET2-PWY</th>\n",
       "      <th>3-HYDROXYPHENYLACETATE-DEGRADATION-PWY</th>\n",
       "      <th>AEROBACTINSYN-PWY</th>\n",
       "      <th>ALL-CHORISMATE-PWY</th>\n",
       "      <th>ANAEROFRUCAT-PWY</th>\n",
       "      <th>ANAGLYCOLYSIS-PWY</th>\n",
       "      <th>ARG+POLYAMINE-SYN</th>\n",
       "      <th>ARGDEG-PWY</th>\n",
       "      <th>ARGORNPROST-PWY</th>\n",
       "      <th>ARGSYN-PWY</th>\n",
       "      <th>...</th>\n",
       "      <th>THISYN-PWY</th>\n",
       "      <th>THREOCAT-PWY</th>\n",
       "      <th>THRESYN-PWY</th>\n",
       "      <th>TRNA-CHARGING-PWY</th>\n",
       "      <th>TRPSYN-PWY</th>\n",
       "      <th>TYRFUMCAT-PWY</th>\n",
       "      <th>UBISYN-PWY</th>\n",
       "      <th>UDPNAGSYN-PWY</th>\n",
       "      <th>VALDEG-PWY</th>\n",
       "      <th>VALSYN-PWY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERR1842970</th>\n",
       "      <td>52346.150689</td>\n",
       "      <td>934.156088</td>\n",
       "      <td>3.040647</td>\n",
       "      <td>3870.324345</td>\n",
       "      <td>42970.456693</td>\n",
       "      <td>63668.502233</td>\n",
       "      <td>21935.119260</td>\n",
       "      <td>226.036347</td>\n",
       "      <td>13264.712099</td>\n",
       "      <td>45083.314734</td>\n",
       "      <td>...</td>\n",
       "      <td>44468.142302</td>\n",
       "      <td>309.349553</td>\n",
       "      <td>62413.129261</td>\n",
       "      <td>61351.902333</td>\n",
       "      <td>47266.719352</td>\n",
       "      <td>4484.484209</td>\n",
       "      <td>8342.892306</td>\n",
       "      <td>40449.436070</td>\n",
       "      <td>143.307666</td>\n",
       "      <td>61824.060934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR1842971</th>\n",
       "      <td>56413.523340</td>\n",
       "      <td>831.370306</td>\n",
       "      <td>20.453622</td>\n",
       "      <td>5508.926323</td>\n",
       "      <td>46847.550535</td>\n",
       "      <td>70713.195125</td>\n",
       "      <td>16966.797283</td>\n",
       "      <td>198.384934</td>\n",
       "      <td>15751.126961</td>\n",
       "      <td>44624.327807</td>\n",
       "      <td>...</td>\n",
       "      <td>42225.970167</td>\n",
       "      <td>314.510945</td>\n",
       "      <td>60791.852851</td>\n",
       "      <td>64375.139723</td>\n",
       "      <td>45748.446330</td>\n",
       "      <td>5938.355727</td>\n",
       "      <td>8033.957521</td>\n",
       "      <td>38659.621301</td>\n",
       "      <td>235.454858</td>\n",
       "      <td>75229.578753</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR1842972</th>\n",
       "      <td>47647.797484</td>\n",
       "      <td>687.881208</td>\n",
       "      <td>411.498246</td>\n",
       "      <td>5500.732680</td>\n",
       "      <td>37796.112232</td>\n",
       "      <td>58733.838610</td>\n",
       "      <td>18927.459528</td>\n",
       "      <td>297.455963</td>\n",
       "      <td>11679.139400</td>\n",
       "      <td>38928.271704</td>\n",
       "      <td>...</td>\n",
       "      <td>43327.432760</td>\n",
       "      <td>1863.888392</td>\n",
       "      <td>57195.214134</td>\n",
       "      <td>56394.345686</td>\n",
       "      <td>43184.453501</td>\n",
       "      <td>2307.532358</td>\n",
       "      <td>6687.211993</td>\n",
       "      <td>35324.842207</td>\n",
       "      <td>280.801577</td>\n",
       "      <td>53983.438560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR1842973</th>\n",
       "      <td>37309.921865</td>\n",
       "      <td>288.217116</td>\n",
       "      <td>0.604144</td>\n",
       "      <td>2008.952399</td>\n",
       "      <td>31496.961837</td>\n",
       "      <td>45345.076618</td>\n",
       "      <td>16486.027854</td>\n",
       "      <td>67.795221</td>\n",
       "      <td>10632.157195</td>\n",
       "      <td>32262.963282</td>\n",
       "      <td>...</td>\n",
       "      <td>32219.964975</td>\n",
       "      <td>244.301131</td>\n",
       "      <td>44779.196097</td>\n",
       "      <td>44065.441393</td>\n",
       "      <td>33630.676807</td>\n",
       "      <td>3431.862477</td>\n",
       "      <td>5238.464062</td>\n",
       "      <td>30786.188559</td>\n",
       "      <td>181.695532</td>\n",
       "      <td>41869.629507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR1842974</th>\n",
       "      <td>36878.822520</td>\n",
       "      <td>353.250362</td>\n",
       "      <td>0.292413</td>\n",
       "      <td>2975.230774</td>\n",
       "      <td>33060.071845</td>\n",
       "      <td>45947.564888</td>\n",
       "      <td>11309.931573</td>\n",
       "      <td>69.752082</td>\n",
       "      <td>10354.793078</td>\n",
       "      <td>25203.534608</td>\n",
       "      <td>...</td>\n",
       "      <td>32195.958904</td>\n",
       "      <td>634.156748</td>\n",
       "      <td>40899.531711</td>\n",
       "      <td>42279.108847</td>\n",
       "      <td>27464.629074</td>\n",
       "      <td>2812.640875</td>\n",
       "      <td>5056.826607</td>\n",
       "      <td>23718.959698</td>\n",
       "      <td>286.003034</td>\n",
       "      <td>38985.336678</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 489 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              1CMET2-PWY  3-HYDROXYPHENYLACETATE-DEGRADATION-PWY  \\\n",
       "ERR1842970  52346.150689                              934.156088   \n",
       "ERR1842971  56413.523340                              831.370306   \n",
       "ERR1842972  47647.797484                              687.881208   \n",
       "ERR1842973  37309.921865                              288.217116   \n",
       "ERR1842974  36878.822520                              353.250362   \n",
       "\n",
       "            AEROBACTINSYN-PWY  ALL-CHORISMATE-PWY  ANAEROFRUCAT-PWY  \\\n",
       "ERR1842970           3.040647         3870.324345      42970.456693   \n",
       "ERR1842971          20.453622         5508.926323      46847.550535   \n",
       "ERR1842972         411.498246         5500.732680      37796.112232   \n",
       "ERR1842973           0.604144         2008.952399      31496.961837   \n",
       "ERR1842974           0.292413         2975.230774      33060.071845   \n",
       "\n",
       "            ANAGLYCOLYSIS-PWY  ARG+POLYAMINE-SYN  ARGDEG-PWY  ARGORNPROST-PWY  \\\n",
       "ERR1842970       63668.502233       21935.119260  226.036347     13264.712099   \n",
       "ERR1842971       70713.195125       16966.797283  198.384934     15751.126961   \n",
       "ERR1842972       58733.838610       18927.459528  297.455963     11679.139400   \n",
       "ERR1842973       45345.076618       16486.027854   67.795221     10632.157195   \n",
       "ERR1842974       45947.564888       11309.931573   69.752082     10354.793078   \n",
       "\n",
       "              ARGSYN-PWY  ...    THISYN-PWY  THREOCAT-PWY   THRESYN-PWY  \\\n",
       "ERR1842970  45083.314734  ...  44468.142302    309.349553  62413.129261   \n",
       "ERR1842971  44624.327807  ...  42225.970167    314.510945  60791.852851   \n",
       "ERR1842972  38928.271704  ...  43327.432760   1863.888392  57195.214134   \n",
       "ERR1842973  32262.963282  ...  32219.964975    244.301131  44779.196097   \n",
       "ERR1842974  25203.534608  ...  32195.958904    634.156748  40899.531711   \n",
       "\n",
       "            TRNA-CHARGING-PWY    TRPSYN-PWY  TYRFUMCAT-PWY   UBISYN-PWY  \\\n",
       "ERR1842970       61351.902333  47266.719352    4484.484209  8342.892306   \n",
       "ERR1842971       64375.139723  45748.446330    5938.355727  8033.957521   \n",
       "ERR1842972       56394.345686  43184.453501    2307.532358  6687.211993   \n",
       "ERR1842973       44065.441393  33630.676807    3431.862477  5238.464062   \n",
       "ERR1842974       42279.108847  27464.629074    2812.640875  5056.826607   \n",
       "\n",
       "            UDPNAGSYN-PWY  VALDEG-PWY    VALSYN-PWY  \n",
       "ERR1842970   40449.436070  143.307666  61824.060934  \n",
       "ERR1842971   38659.621301  235.454858  75229.578753  \n",
       "ERR1842972   35324.842207  280.801577  53983.438560  \n",
       "ERR1842973   30786.188559  181.695532  41869.629507  \n",
       "ERR1842974   23718.959698  286.003034  38985.336678  \n",
       "\n",
       "[5 rows x 489 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pa.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a329bbe",
   "metadata": {},
   "source": [
    "You can see that they look just like the other feature tables we worked before with. The difference is that now they do not contain information about ASVs but about different levels of the functional profiles:\n",
    "\n",
    "1. `ko` table: columns represent KEGG orthologs, as indicated by their names (e.g., **K**19777)\n",
    "2. `ec` table: columns represent enzymes, as indicated by the Enzyme Commission numbers (e.g., **EC**:1.1.1.108)\n",
    "3. `pa` table: columns represent entire pathways using the MetaCyc classification (e.g., ANAGLYCOLYSIS-PWY)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db35c21d",
   "metadata": {},
   "source": [
    "<a id='ipath'></a>\n",
    "### 1.1 KEGG orthologs visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2a0fd440",
   "metadata": {},
   "outputs": [],
   "source": [
    "ko_meta = ko.merge(metadata[['Diets']], left_index=True, right_index=True)\n",
    "ec_meta = ec.merge(metadata[['Diets']], left_index=True, right_index=True)\n",
    "pa_meta = pa.merge(metadata[['Diets']], left_index=True, right_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0a50441",
   "metadata": {},
   "source": [
    "calculating an average abundance of each KO, EC and pathway in each group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "a56f5bfc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# collapse samples per diet types and calculate average abundance\n",
    "\n",
    "ko_meta_avg = ko_meta.groupby('Diets').mean()\n",
    "ec_meta_avg = ec_meta.groupby('Diets').mean()\n",
    "pa_meta_avg = pa_meta.groupby('Diets').mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "36830ebe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>K00001</th>\n",
       "      <th>K00002</th>\n",
       "      <th>K00003</th>\n",
       "      <th>K00004</th>\n",
       "      <th>K00005</th>\n",
       "      <th>K00006</th>\n",
       "      <th>K00007</th>\n",
       "      <th>K00008</th>\n",
       "      <th>K00009</th>\n",
       "      <th>K00010</th>\n",
       "      <th>...</th>\n",
       "      <th>K19777</th>\n",
       "      <th>K19778</th>\n",
       "      <th>K19779</th>\n",
       "      <th>K19780</th>\n",
       "      <th>K19784</th>\n",
       "      <th>K19785</th>\n",
       "      <th>K19787</th>\n",
       "      <th>K19788</th>\n",
       "      <th>K19789</th>\n",
       "      <th>K19791</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Diets</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>corn</th>\n",
       "      <td>13252.028012</td>\n",
       "      <td>884.135704</td>\n",
       "      <td>23651.045340</td>\n",
       "      <td>2260.277845</td>\n",
       "      <td>10733.953733</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>173.282138</td>\n",
       "      <td>14716.065926</td>\n",
       "      <td>2722.162451</td>\n",
       "      <td>5499.498154</td>\n",
       "      <td>...</td>\n",
       "      <td>81.575545</td>\n",
       "      <td>68.853203</td>\n",
       "      <td>3.075969e-01</td>\n",
       "      <td>2.704955e-01</td>\n",
       "      <td>2344.612622</td>\n",
       "      <td>8.142615e-131</td>\n",
       "      <td>0.006193</td>\n",
       "      <td>0.015575</td>\n",
       "      <td>1291.257693</td>\n",
       "      <td>0.048167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>grass</th>\n",
       "      <td>9545.315732</td>\n",
       "      <td>732.287206</td>\n",
       "      <td>21046.649428</td>\n",
       "      <td>1643.740295</td>\n",
       "      <td>7112.936005</td>\n",
       "      <td>0.037495</td>\n",
       "      <td>67.718219</td>\n",
       "      <td>11566.807662</td>\n",
       "      <td>1960.205124</td>\n",
       "      <td>4639.187286</td>\n",
       "      <td>...</td>\n",
       "      <td>16.463126</td>\n",
       "      <td>15.605550</td>\n",
       "      <td>2.484474e-48</td>\n",
       "      <td>2.506367e-48</td>\n",
       "      <td>2107.402170</td>\n",
       "      <td>5.208854e+00</td>\n",
       "      <td>0.001806</td>\n",
       "      <td>0.159915</td>\n",
       "      <td>1245.314649</td>\n",
       "      <td>0.721017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hay</th>\n",
       "      <td>12844.075673</td>\n",
       "      <td>772.263183</td>\n",
       "      <td>22553.639688</td>\n",
       "      <td>1572.156720</td>\n",
       "      <td>12386.494124</td>\n",
       "      <td>0.009775</td>\n",
       "      <td>19.146632</td>\n",
       "      <td>12652.716043</td>\n",
       "      <td>1876.444738</td>\n",
       "      <td>4246.313792</td>\n",
       "      <td>...</td>\n",
       "      <td>9.574571</td>\n",
       "      <td>2.527922</td>\n",
       "      <td>6.299361e-48</td>\n",
       "      <td>4.758332e-02</td>\n",
       "      <td>2328.318036</td>\n",
       "      <td>5.458807e+00</td>\n",
       "      <td>0.005698</td>\n",
       "      <td>0.088389</td>\n",
       "      <td>1364.356658</td>\n",
       "      <td>0.379204</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 10535 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             K00001      K00002        K00003       K00004        K00005  \\\n",
       "Diets                                                                      \n",
       "corn   13252.028012  884.135704  23651.045340  2260.277845  10733.953733   \n",
       "grass   9545.315732  732.287206  21046.649428  1643.740295   7112.936005   \n",
       "hay    12844.075673  772.263183  22553.639688  1572.156720  12386.494124   \n",
       "\n",
       "         K00006      K00007        K00008       K00009       K00010  ...  \\\n",
       "Diets                                                                ...   \n",
       "corn   0.000000  173.282138  14716.065926  2722.162451  5499.498154  ...   \n",
       "grass  0.037495   67.718219  11566.807662  1960.205124  4639.187286  ...   \n",
       "hay    0.009775   19.146632  12652.716043  1876.444738  4246.313792  ...   \n",
       "\n",
       "          K19777     K19778        K19779        K19780       K19784  \\\n",
       "Diets                                                                  \n",
       "corn   81.575545  68.853203  3.075969e-01  2.704955e-01  2344.612622   \n",
       "grass  16.463126  15.605550  2.484474e-48  2.506367e-48  2107.402170   \n",
       "hay     9.574571   2.527922  6.299361e-48  4.758332e-02  2328.318036   \n",
       "\n",
       "              K19785    K19787    K19788       K19789    K19791  \n",
       "Diets                                                            \n",
       "corn   8.142615e-131  0.006193  0.015575  1291.257693  0.048167  \n",
       "grass   5.208854e+00  0.001806  0.159915  1245.314649  0.721017  \n",
       "hay     5.458807e+00  0.005698  0.088389  1364.356658  0.379204  \n",
       "\n",
       "[3 rows x 10535 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ko_meta_avg.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "cea66c06-ddff-4030-8565-b135776fe2e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# collapse samples per diet types and calculate average abundance\n",
    "ko_meta = ko.merge(metadata[['Phase']], left_index=True, right_index=True)\n",
    "ec_meta = ec.merge(metadata[['Phase']], left_index=True, right_index=True)\n",
    "pa_meta = pa.merge(metadata[['Phase']], left_index=True, right_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "40c6effb-201d-42f0-96a1-2056ed472b65",
   "metadata": {},
   "outputs": [],
   "source": [
    "ko_meta_avg = ko_meta.groupby('Phase').mean()\n",
    "ec_meta_avg = ec_meta.groupby('Phase').mean()\n",
    "pa_meta_avg = pa_meta.groupby('Phase').mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "8578d8b5-4c99-4cf6-8745-3b3b976a4cc7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>K00001</th>\n",
       "      <th>K00002</th>\n",
       "      <th>K00003</th>\n",
       "      <th>K00004</th>\n",
       "      <th>K00005</th>\n",
       "      <th>K00006</th>\n",
       "      <th>K00007</th>\n",
       "      <th>K00008</th>\n",
       "      <th>K00009</th>\n",
       "      <th>K00010</th>\n",
       "      <th>...</th>\n",
       "      <th>K19777</th>\n",
       "      <th>K19778</th>\n",
       "      <th>K19779</th>\n",
       "      <th>K19780</th>\n",
       "      <th>K19784</th>\n",
       "      <th>K19785</th>\n",
       "      <th>K19787</th>\n",
       "      <th>K19788</th>\n",
       "      <th>K19789</th>\n",
       "      <th>K19791</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Phase</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>liquid phase</th>\n",
       "      <td>11652.186279</td>\n",
       "      <td>672.049955</td>\n",
       "      <td>21983.024282</td>\n",
       "      <td>1565.336909</td>\n",
       "      <td>9699.963787</td>\n",
       "      <td>0.006596</td>\n",
       "      <td>88.432031</td>\n",
       "      <td>11590.640936</td>\n",
       "      <td>2085.683969</td>\n",
       "      <td>4416.401787</td>\n",
       "      <td>...</td>\n",
       "      <td>30.949923</td>\n",
       "      <td>23.129073</td>\n",
       "      <td>7.726899e-03</td>\n",
       "      <td>1.119377e-01</td>\n",
       "      <td>2180.870825</td>\n",
       "      <td>4.620427</td>\n",
       "      <td>0.005266</td>\n",
       "      <td>0.066704</td>\n",
       "      <td>1280.682089</td>\n",
       "      <td>0.552473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rumen fluid</th>\n",
       "      <td>14392.178230</td>\n",
       "      <td>987.966302</td>\n",
       "      <td>26936.604600</td>\n",
       "      <td>2105.501660</td>\n",
       "      <td>12653.258650</td>\n",
       "      <td>0.001357</td>\n",
       "      <td>103.801933</td>\n",
       "      <td>15655.399135</td>\n",
       "      <td>2970.445427</td>\n",
       "      <td>6502.643506</td>\n",
       "      <td>...</td>\n",
       "      <td>33.276555</td>\n",
       "      <td>21.846664</td>\n",
       "      <td>2.998700e-01</td>\n",
       "      <td>2.061412e-01</td>\n",
       "      <td>2705.157147</td>\n",
       "      <td>3.989805</td>\n",
       "      <td>0.005628</td>\n",
       "      <td>0.078473</td>\n",
       "      <td>1502.000323</td>\n",
       "      <td>0.355185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>solid phase</th>\n",
       "      <td>9597.054908</td>\n",
       "      <td>728.669836</td>\n",
       "      <td>18331.705573</td>\n",
       "      <td>1805.336291</td>\n",
       "      <td>7880.161425</td>\n",
       "      <td>0.039317</td>\n",
       "      <td>67.913025</td>\n",
       "      <td>11689.549560</td>\n",
       "      <td>1502.682917</td>\n",
       "      <td>3465.953939</td>\n",
       "      <td>...</td>\n",
       "      <td>43.386765</td>\n",
       "      <td>42.010938</td>\n",
       "      <td>5.743134e-48</td>\n",
       "      <td>5.793741e-48</td>\n",
       "      <td>1894.304857</td>\n",
       "      <td>2.057429</td>\n",
       "      <td>0.002804</td>\n",
       "      <td>0.118701</td>\n",
       "      <td>1118.246589</td>\n",
       "      <td>0.240730</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 10535 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    K00001      K00002        K00003       K00004  \\\n",
       "Phase                                                               \n",
       "liquid phase  11652.186279  672.049955  21983.024282  1565.336909   \n",
       "rumen fluid   14392.178230  987.966302  26936.604600  2105.501660   \n",
       "solid phase    9597.054908  728.669836  18331.705573  1805.336291   \n",
       "\n",
       "                    K00005    K00006      K00007        K00008       K00009  \\\n",
       "Phase                                                                         \n",
       "liquid phase   9699.963787  0.006596   88.432031  11590.640936  2085.683969   \n",
       "rumen fluid   12653.258650  0.001357  103.801933  15655.399135  2970.445427   \n",
       "solid phase    7880.161425  0.039317   67.913025  11689.549560  1502.682917   \n",
       "\n",
       "                   K00010  ...     K19777     K19778        K19779  \\\n",
       "Phase                      ...                                       \n",
       "liquid phase  4416.401787  ...  30.949923  23.129073  7.726899e-03   \n",
       "rumen fluid   6502.643506  ...  33.276555  21.846664  2.998700e-01   \n",
       "solid phase   3465.953939  ...  43.386765  42.010938  5.743134e-48   \n",
       "\n",
       "                    K19780       K19784    K19785    K19787    K19788  \\\n",
       "Phase                                                                   \n",
       "liquid phase  1.119377e-01  2180.870825  4.620427  0.005266  0.066704   \n",
       "rumen fluid   2.061412e-01  2705.157147  3.989805  0.005628  0.078473   \n",
       "solid phase   5.793741e-48  1894.304857  2.057429  0.002804  0.118701   \n",
       "\n",
       "                   K19789    K19791  \n",
       "Phase                                \n",
       "liquid phase  1280.682089  0.552473  \n",
       "rumen fluid   1502.000323  0.355185  \n",
       "solid phase   1118.246589  0.240730  \n",
       "\n",
       "[3 rows x 10535 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ko_meta_avg.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "0ed3c104",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saving 105 most abundant features...\n",
      "Saving 87 most abundant features...\n",
      "Saving 5 most abundant features...\n"
     ]
    }
   ],
   "source": [
    "# finding the most abundant features in each table\n",
    "# according to the most abundant KOs, ECs and pathways \n",
    "\n",
    "def find_most_abundant(df: pd.DataFrame, frac):\n",
    "    if 0 < frac < 1:\n",
    "        frac = int(frac * len(df.columns))\n",
    "    print(f'Saving {frac} most abundant features...')\n",
    "    most_abundant = {\n",
    "        smp: df.loc[smp, :].sort_values(ascending=False)[:frac]\n",
    "        for smp in df.index\n",
    "    }\n",
    "    return most_abundant\n",
    "\n",
    "ko_most_abundant = find_most_abundant(ko_meta_avg, 0.01)\n",
    "ec_most_abundant = find_most_abundant(ec_meta_avg, 0.03)\n",
    "pa_most_abundant = find_most_abundant(pa_meta_avg, 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20df9047-7231-472a-80b2-01042896f264",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f'5 most abundant KOs in the diets types: {ko_most_abundant[\"yes\"].index[:5].tolist()}\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9dfb00d3",
   "metadata": {},
   "source": [
    "visualizing the KOs and ECs using- iPath 3. After fetching pathway maps for KOs and ECs per sample group and display them as SVGs. Then, producing interactive pathway map available and trying to interpreting different areas of the map."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "855955d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "for smp in ko_most_abundant.keys():\n",
    "    fetch_ipath(ko_most_abundant[smp].index.tolist(), f'{data_dir}/kos_{smp}.svg')\n",
    "    fetch_ipath(ec_most_abundant[smp].index.str.replace(':', '').tolist(), f'{data_dir}/ecs_{smp}.svg')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1aac16b",
   "metadata": {},
   "source": [
    "<a id='metacyc'></a>\n",
    "### 1.2 Enriched pathways\n",
    "\n",
    "The most abundant pathways in the sample groups. we used functions that defined to identify the four pathways that are most abundant in the samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "dad015e4",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'yes'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_3535/1500420391.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf'4 most abundant pathways in the diets group are: {pa_most_abundant[\"yes\"].index[:4].tolist()}\\n'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m: 'yes'"
     ]
    }
   ],
   "source": [
    "print(f'4 most abundant pathways in the diets group are: {pa_most_abundant[\"yes\"].index[:4].tolist()}\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93c3252c",
   "metadata": {},
   "source": [
    "find out differentially abundant features between samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "8c635ea4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[33mQIIME is caching your current deployment for improved performance. This may take a few moments and should only happen once per deployment.\u001b[0m\n",
      "\u001b[32mSaved FeatureTable[Composition] to: Data/picrust2_results/pathway_abundance_bac.qza\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "! qiime composition add-pseudocount \\\n",
    "    --i-table $data_dir/picrust2_results/pathway_abundance.qza \\\n",
    "    --o-composition-table $data_dir/picrust2_results/pathway_abundance_bac.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "b01e4697",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: Data/pa_ancom_abundance_bac.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "! qiime composition ancom \\\n",
    "    --i-table $data_dir/picrust2_results/pathway_abundance_bac.qza \\\n",
    "    --m-metadata-file $data_dir/0-metadata_bac.tsv \\\n",
    "    --m-metadata-column Diets \\\n",
    "    --p-transform-function log \\\n",
    "    --o-visualization $data_dir/pa_ancom_abundance_bac.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "d9a59354",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2-archive-wiaktdf7')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 969239db-a011-4725-8e36-5922393fb88e>"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load(f'{data_dir}/pa_ancom_abundance_bac.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb870a1f-40d7-4fad-945a-add0ee70a6c2",
   "metadata": {},
   "source": [
    "# Functional inference archaea"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "79a59fdf-0785-495d-8bea-e3a4dc37fc6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import qiime2 as q2\n",
    "import requests\n",
    "\n",
    "from qiime2 import Visualization\n",
    "\n",
    "data_dir = 'Data'\n",
    "    \n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "71a583f9-58af-41b4-a14e-f4e4d5234051",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_ipath(ids: list, img_output_path: str, verbose: bool = False):\n",
    "    \"\"\"Fetches a enriched pathways map from iPATH3 for given IDs.\"\"\"\n",
    "    url = 'https://pathways.embl.de/mapping.cgi'\n",
    "    \n",
    "    # remove colon from EC names\n",
    "    if ':' in ids[0]:\n",
    "        ids = [x.replace(':', '') for x in ids]\n",
    "    \n",
    "    if verbose:\n",
    "        print(f'Fetching iPATH3 diagram for ids: {ids}')\n",
    "    params = {\n",
    "        'default_opacity': 0.6,\n",
    "        'export_type': 'svg',\n",
    "        'selection': '\\n'.join(ids)\n",
    "    }   \n",
    "    response = requests.get(url=url, params=params)\n",
    "    \n",
    "    with open(img_output_path, 'wb') as img:\n",
    "        img.write(response.content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "fdbe1d42-1313-451f-a61a-0327ed59843b",
   "metadata": {},
   "outputs": [],
   "source": [
    "picrust_env = '/opt/conda/envs/picrust2/bin'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d4ed0b34-dd6e-41d1-bee9-ea6219e0ea44",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved FeatureTable[Frequency] to: Data/picrust2_results_arc1/ko_metagenome.qza\n",
      "Saved FeatureTable[Frequency] to: Data/picrust2_results_arc1/ec_metagenome.qza\n",
      "Saved FeatureTable[Frequency] to: Data/picrust2_results_arc1/pathway_abundance.qza\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "QIIME is caching your current deployment for improved performance. This may take a few moments and should only happen once per deployment.\n"
     ]
    }
   ],
   "source": [
    "%%script env picrust_env=\"$picrust_env\" data_dir=\"$data_dir\" bash\n",
    "\n",
    "# append the env location to PATH so that qiime\n",
    "# can find all required executables\n",
    "export PATH=$picrust_env:$PATH\n",
    "\n",
    "$picrust_env/qiime picrust2 full-pipeline \\\n",
    "    --i-seq $data_dir/1-rep-seqs_arc.qza \\\n",
    "    --i-table $data_dir/1-feature-table_arc.qza \\\n",
    "    --output-dir $data_dir/picrust2_results_arc1 \\\n",
    "    --p-placement-tool sepp \\\n",
    "    --p-threads 2 \\\n",
    "    --p-hsp-method pic \\\n",
    "    --p-max-nsti 2 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "73a1e122-c41e-44b0-97a3-187abfd4d3f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = pd.read_csv(f'{data_dir}/0-metadata_arc.tsv', sep='\\t', header=0, index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "4b825bd4-0f6c-4227-98fd-5ee2822a6097",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: Data/0-metadata_arc.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "! qiime metadata tabulate \\\n",
    "    --m-input-file $data_dir/0-metadata_arc.tsv \\\n",
    "    --o-visualization $data_dir/0-metadata_arc.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "4a3e01d8-8404-458b-b885-08ae17ade52c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2-archive-xyeha5fq')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 835af58e-226c-4fb7-96d7-48abe8b106d6>"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load(f'{data_dir}/0-metadata_arc.qzv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "60f5218a-7230-489c-a934-d5840f872719",
   "metadata": {},
   "outputs": [],
   "source": [
    "## read all three artifacts using QIIME 2 Python API and view as DaraFrame\n",
    "\n",
    "ko = q2.Artifact.load(f'{data_dir}/picrust2_results_arc/ko_metagenome.qza').view(pd.DataFrame)\n",
    "ec = q2.Artifact.load(f'{data_dir}/picrust2_results_arc/ec_metagenome.qza').view(pd.DataFrame)\n",
    "pa = q2.Artifact.load(f'{data_dir}/picrust2_results_arc/pathway_abundance.qza').view(pd.DataFrame)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "891e36ff-42f1-4f8e-8c79-9df095dc7562",
   "metadata": {},
   "outputs": [],
   "source": [
    "## kegg visualization\n",
    "ko_meta = ko.merge(metadata[['Diets']], left_index=True, right_index=True)\n",
    "ec_meta = ec.merge(metadata[['Diets']], left_index=True, right_index=True)\n",
    "pa_meta = pa.merge(metadata[['Diets']], left_index=True, right_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "4a0317fc-0308-443f-acf8-e0a1d5fbbbd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# collapse samples per diet types and calculate average abundance\n",
    "\n",
    "ko_meta_avg = ko_meta.groupby('Diets').mean()\n",
    "ec_meta_avg = ec_meta.groupby('Diets').mean()\n",
    "pa_meta_avg = pa_meta.groupby('Diets').mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "ffedbd8d-3db1-4463-a205-79b5fb0838b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# collapse samples per diet types and calculate average abundance\n",
    "ko_meta = ko.merge(metadata[['Phase']], left_index=True, right_index=True)\n",
    "ec_meta = ec.merge(metadata[['Phase']], left_index=True, right_index=True)\n",
    "pa_meta = pa.merge(metadata[['Phase']], left_index=True, right_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "2a9942bf-eca1-4ee7-a823-08c5b6adb0b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "ko_meta_avg = ko_meta.groupby('Phase').mean()\n",
    "ec_meta_avg = ec_meta.groupby('Phase').mean()\n",
    "pa_meta_avg = pa_meta.groupby('Phase').mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "f221ce66-9e89-4c7e-9295-6ec48ba7d700",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saving 93 most abundant features...\n",
      "Saving 82 most abundant features...\n",
      "Saving 5 most abundant features...\n"
     ]
    }
   ],
   "source": [
    "# finding the most abundant features in each table\n",
    "# according to the most abundant KOs, ECs and pathways \n",
    "\n",
    "def find_most_abundant(df: pd.DataFrame, frac):\n",
    "    if 0 < frac < 1:\n",
    "        frac = int(frac * len(df.columns))\n",
    "    print(f'Saving {frac} most abundant features...')\n",
    "    most_abundant = {\n",
    "        smp: df.loc[smp, :].sort_values(ascending=False)[:frac]\n",
    "        for smp in df.index\n",
    "    }\n",
    "    return most_abundant\n",
    "\n",
    "ko_most_abundant = find_most_abundant(ko_meta_avg, 0.01)\n",
    "ec_most_abundant = find_most_abundant(ec_meta_avg, 0.03)\n",
    "pa_most_abundant = find_most_abundant(pa_meta_avg, 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "4aabd42f-bf1f-479f-b79f-6f503c729290",
   "metadata": {},
   "outputs": [],
   "source": [
    "## fetching pathways and visualization\n",
    "\n",
    "for smp in ko_most_abundant.keys():\n",
    "    fetch_ipath(ko_most_abundant[smp].index.tolist(), f'{data_dir}/kos_{smp}.svg')\n",
    "    fetch_ipath(ec_most_abundant[smp].index.str.replace(':', '').tolist(), f'{data_dir}/ecs_{smp}.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "ce113713-2cb7-436b-b3f5-21c0891411b0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureTable[Composition] to: Data/picrust2_results_arc/pathway_abundance_arc.qza\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "! qiime composition add-pseudocount \\\n",
    "    --i-table $data_dir/picrust2_results_arc/pathway_abundance.qza \\\n",
    "    --o-composition-table $data_dir/picrust2_results_arc/pathway_abundance_arc.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "bcc862b2-fe2f-461d-8a39-2dcdfff866f5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: Data/pa_ancom_abundance_arc.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "! qiime composition ancom \\\n",
    "    --i-table $data_dir/picrust2_results_arc/pathway_abundance_arc.qza \\\n",
    "    --m-metadata-file $data_dir/0-metadata_arc.tsv \\\n",
    "    --m-metadata-column Diets \\\n",
    "    --p-transform-function log \\\n",
    "    --o-visualization $data_dir/pa_ancom_abundance_arc.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "99e5a735-d616-45fd-b64c-1fdb2485d848",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2-archive-x3aqivgm')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 8dc6b31a-1e81-4830-9203-0bb1f65780a7>"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load(f'{data_dir}/pa_ancom_abundance_arc.qzv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2e5ee6f-b662-4d8c-91f4-aa72b92782d1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "QIIME2 (picrust2)",
   "language": "python",
   "name": "picrust2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
